arm2hpdl.1
'\" t
'\"
'\"	DO NOT EDIT! This file is generated from arm2hpdl.1in
'\"
.TH arm2hpdl 1 "Sun Apr 25 16:22:30 2021" "arm2hpdl 0.0"
'\"
'\"==========================================================================
'\"	STRINGS and MACROS
'\"==========================================================================
'\"
'\"	Define strings for special characters that nroff doesn't have
'\"
'\"	N.B. using .if cua to test for special characters did not work.
'\"
.if !'\*[.T]'ps' .ds ua up
.if  '\*[.T]'ps' .ds ua \(ua
.if !'\*[.T]'ps' .ds da down
.if  '\*[.T]'ps' .ds da \(da
.if !'\*[.T]'ps' .ds <- left
.if  '\*[.T]'ps' .ds <- \(<-
.if !'\*[.T]'ps' .ds -> right
.if  '\*[.T]'ps' .ds -> \(->
'\"
'\"     bx - box a Courier string for making keycaps
'\"
'\"	N.B. this mess is to make the line drawing come out only
'\"	if we are really generating postscript
'\"
.de bx
.ie !'\*[.T]'ps' \{\
.	RB [ \\$1 ]\\$2
.\}
.el \{\
.	ie !r ps4html \{\
'\" \(br\|\s-1\f(CB\\$1\fP\s+1\|\(br\l'|0\(rn'\l'|0\(ul'
.		ft CW
.		nr par*bxw \w'\\$1'+.4m 
\Z'\v'.25m'\D'l 0 -1m'\D'l \\n[par*bxw]u 0'\D'l 0 1m'\D'l -\\n[par*bxw]u 0''\
\Z'\h'.2m'\s-1\\$1\s+1'\
\h'\\n[par*bxw]u'\\$2
.		ft P
.	\}
.	el \{\
.		RB [ \\$1 ]\\$2
.	\}
.\}
..
'\"
'\" strings to set current color (null with old groff)
'\"
.if mred .ds red \m[red]
.if mgreen .ds green \m[green]
.if mblue .ds blue \m[blue]
.if mblack .ds black \m[black]
.if mblack .ds mP \mP
'\"
'\" fix for grotty + xterm. We call for orange, grotty outputs yellow,
'\" but xterm displays yellow as orange.  The cycle is complete.
'\"
.if n .defcolor orange rgb #ffff00
'\"
'\" color <color> - set the current color (ignores request with old groff)
'\"
.de color
.if mred \m[\\$1]\c
..
'\"
'\" colorword <color> <word> - colorize a word (ignored by old groff)
'\"
.de colorword
.ie m\\$1 \m[\\$1]\\$2\mP\c
.el \\$2\c
..
'\"
'\" colbox <fg> <bg> <word> - colorize a word in a filled box
'\"
.de colbox
.ie mred \M[\\$2]\
\v'+.167v'\
\D'P 0 -0.9v  \w'\\$3'u 0  0 +0.9v   -\w'\\$3'u 0'\
\v'-.167v'\
\m[\\$1]\\$3\mP\MP
.el \\$3\c
..
'\"
'\"	Macros for doing pdfmarks
'\"
.de specialps
.if  '\*[.T]'ps' \\k_\X'ps: \\$*'\h'|\\n_u'\c
..
'\"
'\" pdfmark PDFMARKCODE
'\"
.ds pdfmarks
.if d pdfmarks \{\
.de pdfmark
.	specialps exec [\\$1 pdfmark
..
'\"
'\" pdfdest LINKNAME
'\"
.de pdfdest
.pdfmark "/Dest /\\$1 /View [/XYZ -5 PL null] /DEST"
..
'\"
'\" pdfbookmark COUNT LINKNAME STRING
'\"
.de pdfbookmark
.   pdfmark "/View [/XYZ 44 730 1.0] /Count \\$1 /Dest /\\$2 /Title (\\$3) /OUT"
..
'\"
'\"	Define the SH and SS macros to save pdfmark information
'\"	in "arrays" of numbers and strings.
'\"
.if !r rr_n \{\
.nr rr_n 0 1
.am SH
.	nr rr_levels!\\n+[rr_n] 2
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.am SS
.	nr rr_levels!\\n+[rr_n] 3
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.\}
'\"
'\"	Called at the end of the document to generate the pdfmark outline
'\"
.de pdf_outline
.nr rr_levels!\\n+[rr_n] 1
.nr rr_i 0 1
.while \\n+[rr_i]<\\n[rr_n] \{\
.   nr rr_ip1 \\n[rr_i]+1
.   nr rr_count 0
.   if \\n[rr_levels!\\n[rr_ip1]]>\\n[rr_levels!\\n[rr_i]] \{\
.       nr rr_j \\n[rr_i] 1
.       while \\n+[rr_j]<\\n[rr_n] \{\
.           if \\n[rr_levels!\\n[rr_j]]<=\\n[rr_levels!\\n[rr_i]] \{\
.               break
.           \}
.           if \\n[rr_levels!\\n[rr_j]]==(\\n[rr_levels!\\n[rr_i]]+1) \{\
.               nr rr_count \\n[rr_count]+1
.           \}
.       \}
.   \}
.   ds hhh \\*[rr_labels!\\n[rr_i]]
.   pdfbookmark -\\n[rr_count] Link\\n[rr_i] "\\*[hhh]"
.\}
..
'\"
'\" Some postscript to make pdfmarks harmless on old interpreters...
'\"
.specialps "def /pdfmark where {pop} {userdict /pdfmark /cleartomark load put} ifelse"
'\"
'\" Force display of Bookmarks in Acrobat when document is viewed.
'\"
.pdfmark "[/PageMode /UseOutlines /Page 1 /View [/XYZ null null null] /DOCVIEW"
'\"
'\" Output the document info in pdfmarks
'\"
.pdfmark "\
	/Title (\*[an-title](\*[an-section])) \
	/Subject (\*[an-title] Manual Page) \
	/Author (Rick Richardson) \
	/Keywords (printing) \
	/Creator (groff \n(.x.\n(.y.\n(.Y -man) \
	/CreationDate (\*[an-extra1]) \
	/ModDate (\*[an-extra1]) \
	/DOCINFO"
\}
'\"
'\" The manual page name is only 1st level mark
'\"
.nr rr_levels!\n+[rr_n] 1
.ds rr_labels!\n[rr_n] \*[an-title](\*[an-section])
.pdfdest Link\n[rr_n]
'\"
'\"==========================================================================
'\"	MANUAL PAGE SOURCE
'\"==========================================================================
.SH NAME
arm2hpdl \- Add HP download header/trailer to an ARM ELF binary.
.SH SYNOPSIS
.B arm2hpdl
.RI [ options "] arm-binary.img > hpdl.dl"
.SH DESCRIPTION
.B arm2hpdl
adds an HP download header/trailer to an ARM ELF binary.
If the file already has an HP header, just copy it to stdout.
.SH COMMAND LINE OPTIONS
These are the options that can appear on the command line.
.TP
.BI \-D\0 level
Set Debug level [0].

.SH EXAMPLES
Add an HPDL header to a HP LaserJet 1005.

.nf
.ft CW
$ arm2hpdl sihp1005.img > sihp1005.dl
.ft P
.fi

.SH FILES
.BR /usr/bin/arm2hpdl,
.BR /usr/share/foo2*/firmware/
.SH SEE ALSO
.BR foo2zjs (1)
.SH "AUTHOR"
Rick Richardson <rick.richardson@comcast.net>
.br
http://foo2zjs.rkkda.com/
'\"
'\"
'\"
.em pdf_outline


ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default
	* hbpldecode: Print out paper codes.

2014-03-15	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More changes from Dave.  Still gets 
	    Segmentation fault (core dumped)
	* hbpldecode: Changed unsigned to int in 2 places.  Works good!!!

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Bug with no -d/-r.

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Changes to pick apart version 1 from
	    Dave Coffin <dcoffin@cybercom.net>
	* I get:
	    Decoding page 1 to page-01.ppm ...
	    Segmentation fault (core dumped)

2014-03-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint M215 MFP using the foo2hbpl2 driver.

2014-02-09	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox WorkCentre 3045 MFP using the foo2hbpl2 driver.

2014-01-26	Rick Richardson <rick.richardson@comcast.net>
	* foo2qpdl et al: with -z3 (e.g. CLP-365) you can adjust
	    Brightness and Contrast for monochrome prints, and you can adjust
	    Saturation, Cyan-Red, Magenta-Green, Yellow-Blue Balances.
	* foo2hp: somewhere a bug in foomatic (probably 4.x) crept in
	    so foo2hp-AlignCMYK.xml was only processing one option instead
	    of three options!  Split the file into foo2hp-AlignC.xml,
	    foo2hp-AlignM.xml, and foo2hp-AlignY.xml.

2014-01-24	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Dell C1765nf/nfw using the foo2hbpl2 driver.

2014-01-08	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser M1400 using the foo2hbpl2 driver.

2013-12-25	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: HP LaserJet Pro M1212nf MFP using the foo2xqx driver.

2013-12-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc: clean up man pages

2013-11-29	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc-wrapper, foomatic and PPD stuff:
	    add -z1 for Oki C810 so everything will be

arm2hpdl.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

/*
 * Global option flags
 */
int	Debug = 0;

void
debug(int level, char *fmt, ...)
{
	va_list ap;

	if (Debug < level)
		return;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

int
error(int fatal, char *fmt, ...)
{
	va_list ap;

	if (fatal)
	    fprintf(stderr, "Error: ");
	else
	    fprintf(stderr, "Warning: ");
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);

	if (fatal > 0)
	    exit(fatal);
	else
	    return (fatal);
}

void
usage(void)
{
	fprintf(stderr,
"Usage:\n"
"	arm2hpdl [options] sihp1005.img > sihp1005.dl\n"
"\n"
"	Add HP download header/trailer to an ARM ELF binary.\n"
"	If the file already has an HP header, just copy it to stdout.\n"
"\n"
"Options:\n"
"       -D lvl      Set Debug level [%d]\n"
	, Debug
	);

	exit(1);
}

/*
 * Compute HP-style checksum
 */
long
docheck(long check, unsigned char *buf, int len)
{
    int	i;

    if (len & 1)
	error(1, "We should never see an odd number of bytes in this app.\n");

    for (i = 0; i < len; i += 2)
	check += (buf[i]<<0) | (buf[i+1]<<8);
    return check;
}

int
main(int argc, char *argv[])
{
	extern int	optind;
	extern char	*optarg;
	int		c;
	int		rc;
	unsigned char	buf[BUFSIZ];
	int		len;
	FILE		*fp;
	struct stat	stats;
	int		size;
	unsigned char	elf[4];
	long		check;
	int		iself;
	int		ispjl;

	while ( (c = getopt(argc, argv, "D:?h")) != EOF)
		switch (c)
		{
		case 'D':
			Debug = atoi(optarg);
			break;
		default:
			usage();
			exit(1);
		}

	argc -= optind;
	argv += optind;

	if (argc != 1)
	    usage();

	/*
	 * Open the file and figure out if its an ELF file
	 * by reading the first 4 bytes.
	 */
	fp = fopen(argv[0], "r");
	if (!fp)
	    error(1, "Can't open '%s'\n", argv[0]);

	len = fread(elf, 1, sizeof(elf), fp);
	if (len != 4)
	    error(1, "Premature EOF on '%s'\n", argv[0]);

	iself = 0;
	ispjl = 0;
	check = 0;
	if (memcmp(elf, "\177ELF", 4) == 0)
	{
	    /*
	     * Its an ELF executable file
	     */
	    unsigned char	filhdr[17] =
	    {
		0xbe, 0xef, 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
		0, 0, 0, 0,	/* size goes here */
		0, 0, 0,
	    };
	    unsigned char	sechdr[12] =
	    {
		0xc0, 0xde, 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 0, 0
	    };

	    iself = 1;

	    /*
	     * Create and write the file header
	     */
	    rc = stat(argv[0], &stats);
	    if (rc < 0)
		error(1, "Can't stat '%s'\n", argv[0]);

	    size = stats.st_size + 12 + 4;

	    filhdr[10] = size>>24;
	    filhdr[11] = size>>16;
	    filhdr[12] = size>> 8;
	    filhdr[13] = size>> 0;

	    rc = fwrite(filhdr, 1, sizeof(filhdr), stdout);

	    /*
	     * Create and write the section header
	     */
	    //memset(sechdr+2, 0, sizeof(sechdr)-2);

	    check = docheck(check, sechdr, sizeof(sechdr));
	    rc = fwrite(sechdr, 1, sizeof(sechdr), stdout);
	}
	else if (memcmp(elf, "\276\357AB", 4) == 0)
	{
	    /*
	     * This file already has an HP download header.
	     * Don't change it.
	     */
	    if (Debug)
		error(0, "This file already has an HP header.  "
		    "I will just copy it to stdout.\n");
	}
	else if (memcmp(elf, "20", 2) == 0)
	{
	    unsigned char hdr[8];
	    
	    ispjl = 1;
	    printf("\033%%-12345X@PJL ENTER LANGUAGE=ACL\r\n");

	    rc = stat(argv[0], &stats);
	    if (rc < 0)
		error(1, "Can't stat '%s'\n", argv[0]);

	    size = stats.st_size - 8;

	    hdr[0] = 0x00;
	    hdr[1] = 0xac;
	    hdr[2] = 0xc0;
	    hdr[3] = 0xde;
	    hdr[4] = size>>24;
	    hdr[5] = size>>16;
	    hdr[6] = size>> 8;
	    hdr[7] = size>> 0;

	    rc = fwrite(hdr, 1, sizeof(hdr), stdout);
	}
	else
	{
	    error(1, "I don't understand this file at all!\n");
	}

	/*
	 * Write out the 4 bytes we read earlier
	 */
	if (iself)
	    check = docheck(check, elf, sizeof(elf));
	rc = fwrite(elf, 1, sizeof(elf), stdout);

	/*
	 * Write out the remainder of the file
	 */
	while ( (len = fread(buf, 1, sizeof(buf), fp)) )
	{
	    if (iself)
		check = docheck(check, buf, len);
	    rc = fwrite(buf, 1, len, stdout);
	}

	fclose(fp);

	/*
	 * Add the file trailer
	 */
	if (iself)
	{
	    /*
	     * Add in the checksum carries and complement it
	     */
	    while (check >> 16)
		check = (check&0xffff) + (check>>16);
	    check = ~check;

	    putchar(0xff);
	    putchar(0xff);
	    putchar((check >> 0) & 0xff);
	    putchar((check >> 8) & 0xff);
	    debug(1, "checksum = %lx\n", check);
	}
	if (ispjl)
	    printf("\033%%-12345X");

	exit(0);
}
